File: pydss/common/FeederImpl.py
from dss.common.Feeder import (Feeder,)
from dss.conversion.impl.PCClassImpl import (PCClassImpl,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.shared.impl.CommandListImpl import (CommandListImpl,)
from dss.common.impl.DSSClassDefs import (DSSClassDefs,)
from dss.common.impl.FeederObjImpl import (FeederObjImpl,)
from dss.parser.impl.Parser import (Parser,)


class FeederImpl(PCClassImpl, Feeder):
    NumPropsThisClass = 0
    activeFeederObj = None

    def __init__(self):
        super(FeederImpl, self)()
        self.className = 'Feeder'
        self.classType = DSSClassDefs.FEEDER_ELEMENT
        # + PC_ELEMENT;
        # add to PCElement list
        self.activeElement = -1
        self.defineProperties()
        commands = [None] * self.numProperties
        System.arraycopy(self.propertyName, 0, commands, 0, self.numProperties)
        self.commandList = CommandListImpl(commands)
        self.commandList.setAbbrevAllowed(True)

    def defineProperties(self):
        self.NumPropsThisClass = 0
        self.numProperties = self.NumPropsThisClass
        self.countProperties()
        # get inherited property count
        self.allocatePropertyArrays()
        # Can't think of any properties we want the user to be able to set
        # Define Property names
        # PropertyName[0] = "bus1";
        # define Property help values
        # PropertyHelp[0] = "Name of bus to which source is connected."+DSSGlobals.CRLF+"bus1=busname"+DSSGlobals.CRLF+"bus1=busname.1.2.3";
        self.activeProperty = self.NumPropsThisClass - 1
        super(FeederImpl, self).defineProperties()
        # Add defs of inherited properties to bottom of list

    def newObject(self, objName):
        """Called from EnergyMeter."""
        # make a new feeder object
        # first see if this one already exists; if so, just reinitialize
        obj = self.find(objName)
        ckt = DSSGlobals.activeCircuit
        if obj is not None:
            ckt.setActiveCktElement(obj)
            result = 0
        else:
            ckt.setActiveCktElement(FeederObjImpl(self, objName))
            result = self.addObjectToList(DSSGlobals.activeDSSObject)
            ckt.addCktElement(result)
            # done here because feeder objects are instantiated from energy meters
        return result

    def edit(self):
        # continue parsing with contents of parser
        self.activeFeederObj = self.elementList.getActive()
        DSSGlobals.activeCircuit.setActiveCktElement(self.activeFeederObj)
        result = 0
        paramPointer = 0
        paramName = Parser.getInstance().getNextParam()
        param = Parser.getInstance().makeString()
        while len(param) > 0:
            if len(paramName) == 0:
                paramPointer += 1
            else:
                paramPointer = self.commandList.getCommand(paramName)
            if paramPointer > 0 and paramPointer <= self.numProperties:
                self.activeFeederObj.setPropertyValue(paramPointer, param)
            _0 = paramPointer
            _1 = False
            while True:
                if _0 == 0:
                    _1 = True
                    DSSGlobals.doSimpleMsg('Unknown parameter \"' + paramName + '\" for object \"' + self.getName() + '.' + self.activeFeederObj.getName() + '\"', 630)
                    break
                if True:
                    _1 = True
                    self.classEdit(self.activeFeederObj, paramPointer - self.NumPropsThisClass)
                    break
                break
            paramName = Parser.getInstance().getNextParam()
            param = Parser.getInstance().makeString()
        self.activeFeederObj.recalcElementData()
        self.activeFeederObj.setYPrimInvalid(True)
        return result

    def makeLike(self, otherFeederName):
        result = 0
        # See if we can find this name in the present collection
        otherFeeder = self.find(otherFeederName)
        if otherFeeder is not None:
            if self.activeFeederObj.getNPhases() != otherFeeder.getNPhases():
                self.activeFeederObj.setNPhases(otherFeeder.getNPhases())
                self.activeFeederObj.setNConds(self.activeFeederObj.getNPhases())
                # forces reallocation of terminal stuff
                self.activeFeederObj.setYorder(self.activeFeederObj.getNConds() * self.activeFeederObj.getNTerms())
                self.activeFeederObj.setYPrimInvalid(True)
            # put properties to copy here
            self.classMakeLike(otherFeeder)
            # set spectrum, base frequency
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.activeFeederObj.getParentClass().getNumProperties()):
                    break
                self.activeFeederObj.setPropertyValue(i, otherFeeder.getPropertyValue(i))
            result = 1
        else:
            DSSGlobals.doSimpleMsg('Error in Feeder makeLike: \"' + otherFeederName + '\" not found.', 631)
        return result

    def init(self, handle):
        DSSGlobals.doSimpleMsg('Need to implement Feeder.init()', -1)
        return 0

File: core/dss-core/src/main/java/com/ncond/dss/common/impl/FeederImpl.java
package com.ncond.dss.common.impl;

import com.ncond.dss.common.Circuit;
import com.ncond.dss.common.Feeder;
import com.ncond.dss.common.FeederObj;
import com.ncond.dss.conversion.impl.PCClassImpl;
import com.ncond.dss.parser.impl.Parser;
import com.ncond.dss.shared.impl.CommandListImpl;

public class FeederImpl extends PCClassImpl implements Feeder {

	public static int NumPropsThisClass = 0;

	public static FeederObj activeFeederObj;

	public FeederImpl() {
		super();
		className = "Feeder";
		classType = DSSClassDefs.FEEDER_ELEMENT; /*+ PC_ELEMENT; */ // add to PCElement list

		activeElement = -1;

		defineProperties();

		String[] commands = new String[numProperties];
		System.arraycopy(propertyName, 0, commands, 0, numProperties);
		commandList = new CommandListImpl(commands);
		commandList.setAbbrevAllowed(true);
	}

	protected void defineProperties() {
		NumPropsThisClass = 0;

		numProperties = NumPropsThisClass;
		countProperties();   // get inherited property count
		allocatePropertyArrays();

		// no properties we want the user to be able to set

		// define property names
		//propertyName[0] = "bus1";

		// define Property help values
		//propertyHelp[0] = "Name of bus to which source is connected."+DSSGlobals.CRLF+"bus1=busname"+DSSGlobals.CRLF+"bus1=busname.1.2.3";

		activeProperty = NumPropsThisClass - 1;
		super.defineProperties();  // add defs of inherited properties to bottom of list
	}

	/**
	 * Called from EnergyMeter.
	 */
	@Override
	public int newObject(String objName) {
		int result;

		// make a new feeder object
		// first see if this one already exists; if so, just reinitialize
		FeederObj obj = (FeederObj) find(objName);

		Circuit ckt = DSSGlobals.activeCircuit;

		if (obj != null) {
			ckt.setActiveCktElement((DSSCktElement) obj);
			result = -1;
		} else {
			ckt.setActiveCktElement(new FeederObjImpl(this, objName));
			result = addObjectToList(DSSGlobals.activeDSSObject);
			ckt.addCktElement(result);
			// done here because feeder objects are instantiated from energy meters
		}

		return result;
	}

	@Override
	public int edit() {
		// continue parsing with contents of parser
		activeFeederObj = (FeederObj) elementList.getActive();
		DSSGlobals.activeCircuit.setActiveCktElement((DSSCktElement) activeFeederObj);

		int result = 0;

		int paramPointer = -1;
		String paramName = Parser.getInstance().getNextParam();
		String param = Parser.getInstance().makeString();
		while (param.length() > 0) {
			if (paramName.length() == 0) {
				paramPointer += 1;
			} else {
				paramPointer = commandList.getCommand(paramName);
			}

			if (paramPointer >= 0 && paramPointer < numProperties)
				activeFeederObj.setPropertyValue(paramPointer, param);

			switch (paramPointer) {
			case -1:
				DSSGlobals.doSimpleMsg("Unknown parameter \"" + paramName + "\" for object \"" + getName() +"."+ activeFeederObj.getName() + "\"", 630);
				break;
			default:
				classEdit(activeFeederObj, paramPointer - NumPropsThisClass);
				break;
			}

			paramName = Parser.getInstance().getNextParam();
			param     = Parser.getInstance().makeString();
		}

		activeFeederObj.recalcElementData();
		activeFeederObj.setYPrimInvalid(true);

		return result;
	}

	@Override
	protected int makeLike(String otherFeederName) {
		int result = 0;

		/* See if we can find this name in the present collection */
		FeederObj otherFeeder = (FeederObj) find(otherFeederName);
		if (otherFeeder != null) {
			if (activeFeederObj.getNPhases() != otherFeeder.getNPhases()) {
				activeFeederObj.setNPhases(otherFeeder.getNPhases());
				activeFeederObj.setNConds(activeFeederObj.getNPhases());  // forces reallocation of terminal stuff

				activeFeederObj.setYOrder(activeFeederObj.getNConds() * activeFeederObj.getNTerms());
				activeFeederObj.setYPrimInvalid(true);
			}

			// put properties to copy here

			classMakeLike(otherFeeder);  // set spectrum, base frequency

			for (int i = 0; i < activeFeederObj.getParentClass().getNumProperties(); i++) {
				activeFeederObj.setPropertyValue(i, otherFeeder.getPropertyValue(i));
			}
			result = 1;
		} else {
			DSSGlobals.doSimpleMsg("Error in Feeder makeLike: \"" + otherFeederName + "\" not found.", 631);
		}

		return result;
	}

	@Override
	public int init(int handle) {
		DSSGlobals.doSimpleMsg("Need to implement Feeder.init()", -1);
		return 0;
	}

}
